test/node-api/test_cannot_run_js/test_cannot_run_js.c

```diff
@@ -0,0 +1,46 @@
+#include <node_api.h>
+#include "../../js-native-api/common.h"
+#include "stdlib.h"
+
+static void Finalize(napi_env env, void* data, void* hint) {
+  napi_value cb;
+  napi_ref* ref = data;
+#ifdef NAPI_EXPERIMENTAL
+  napi_status expected_status = napi_cannot_run_js;
+#else
+  napi_status expected_status = napi_pending_exception;
+#endif  // NAPI_EXPERIMENTAL
+
+  if (napi_get_reference_value(env, *ref, &cb) != napi_ok) abort();
```

Comment:
It is not a good practice to get the referenced value of a `napi_ref` in its finalizer. This particular case works because the finalizer is called by the node-api implementation rather than the garbage collector. I don't think this semantic is guaranteed to be stable.

IMO, it would be better to retrieve the global object with `napi_get_global` and access properties on it with `napi_get_property`, which also calls into js, in the finalizer.